rfcs/0011-streaming-cdc.md

```diff
@@ -0,0 +1,188 @@
+to stream data (both production and consumption) because of an object store's
+bottomless storage capacity, ease of operation, and cost-effectiveness.
+
+A streaming consumer API will allow applications to write key-value pairs to
+SlateDB, where the key is a unique identifier for the event and the value is
+the event data. UUIDs, auto-incrementing integers, or even PostgreSQL LSNs are
+all reasonable choices for the key.
+
+Users might also use non-unique keys for their messages, which will cause
+SlateDB to behave like a compacted topic in Kafka. If a user wishes to emulate
+multiple topics, they can do so by:
+
+1. Creating one SlateDB database per topic
+2. Using a prefix in the key to identify the topic
+
+(1) prevents cross-topic transactions, while (2) will require applications to
+filter events from unrelated topics, since they will see rows for all topics in
+the WAL.
+
+_TODO we should expose seqnum so users can provide both key and seqnum to get
+compacted topics like Kafka_
+
+## Current State
+
+SlateDB does not currently support either CDC or messaging. Users that wish to
+receive near-realtime updates to their SlateDB have the following uptions:
+
+1. Poll the database for changes
+2. Dual write to SlateDB and an external message queue
+3. Implement their own CDC solution using SlateDB's WALs
+
+(1) and (2) are not optimal designs. (3) is the correct architecture, but it
+is not a simple implementation. This RFC proposes that we implement (3) in
+SlateDB, so users can stream changes out of SlateDB without additional
+infrastructure.
+
+## Goals
+
+1. Provide a durable, ordered stream of all mutations to a SlateDB database.
+2. Allow clients to resume from a previous position in the stream by seqnum,
+   logical timestamp, or system time.
+3. Prevent dataloss if consumers fall behind recently written WALs.
+4. At-least-once delivery semantics for consumers.
+5. Single-digit second latency for consumers.
+6. WAL SST boundaries. Exposing WAL SST boundaries allows clients to treat the
+   entire WAL SST as a single transaction. This is useful for clients that
+   need to respect transaction boundaries.
+7. Strong ordering. Events are ordered by their seqnum and timestamp.
+8. Allow clients to bootstrap the entire database. Bootstrapping should
+   transparently transition to realtime updates from the WAL.
+
+## Non-Goals
+
+1. Exactly once delivery semantics. Consumers are expected to handle retries.
+2. Event transformations. Streaming will only expose raw WAL rows and SST
+   boundaries.
+3. Streaming from parent DBs or external DBs. Users will only be able to stream
+   clone DBs as far back as the WALs that were copied during clone
+   initialization.
+4. TTL-based row-expiration events. Consumers must track TTL expiration and delete
+   expired data outside of SlateDB.
+5. Update events [similar to Debezium's](https://debezium.io/documentation/reference/stable/transformations/event-changes.html#_change_event_structure).
```

Comment:
we need to figure out how we want to support the merge operation for CDC -- cc @dmvk